### ISSUE TITLE
BUG: np.dtype(ctype) does not respect endianness

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#10533]
```python
class Foo(ctypes.BigEndianStructure):
    _fields_ = [('one', ctypes.c_uint32)]
```
Gives a little-endian result:
```
>>> np.dtype(Foo)
dtype([('one', '<u4')], align=True)
```

**Comment by eric-wieser** [comment]
A nice solution to both this and #10532 would be to simply implement
```
def ctype_to_dtype(ctype):
    len0_arr = (ctype*0)()
    return np.array(memoryview(len0_arr)).dtype
```

Unfortunately, bugs in CPython prevent this from working

* memoryview(ctype_obj) is broken on structs: https://bugs.python.org/issue32780 (python/cpython#5561)
* len(arr) == 0 gives an invalid itemsize: https://bugs.python.org/issue32782 (python/cpython#5576)

**Comment by eric-wieser** [comment]
Fix belongs in these lines:

https://github.com/numpy/numpy/blob/44810f07ec6a281d96d6a69c105ce6ae59aec718/numpy/core/_dtype_ctypes.py#L64-L65

**Comment by bmakos** [comment]
I'll give a try fixing this over the next weeks if nobody else is working on it. I will have to read up a bit on the issue.

**Comment by bmakos** [comment]
Ok, I looked a bit at this. After some testing I arrived at a solution Eric already touched in a [SO question](https://stackoverflow.com/questions/48650192/how-can-i-tell-if-a-ctypes-integer-type-is-big-endian). 

He threw away the solution because some simple types like `ctypes.c_uint8.__ctype_be__` do not have  `__name__.endswith('_be') = true`. Indeed this would be a deal breaker, but then I asked why could this be? Ran a small test: 
```
import sys, unittest, struct, math, ctypes
from binascii import hexlify
from ctypes import *
def bin(s):
    return hexlify(memoryview(s)).decode().upper()
x1 = ctypes.c_int8.__ctype_be__
xb = x1(0x12)
x2 = ctypes.c_int8.__ctype_le__
xl = x2(0x12)

bin(xb)
>>> '12'
bine(xl)
>>>'12'
```

Hmm, so it may look that `c_int8` does not support the change of endianess and this is why it would not report it in `__name__`. 

So with all this in mind, wrote some code that would work with my previous pull where I added _pack_ support and it looks promising. I will do a pull request with the code I think should be added + enabled the tests for this and added a few extra.
